Add OAuth 2.0 authentication and log access tokens (#2951)

nadove-ucsc opened the pull request:
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [x] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [x] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [x] Verified that `N reviews` labelling is accurate
- [x] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [x] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [x] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [x] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [x] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [x] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [x] Unassigned PR

nadove-ucsc committed:
[u R 1/3] Add OAuth 2.0 authentication and log access tokens (#2951)

noah-aviel-dove commented:
Latency results for `https://www.googleapis.com/oauth2/v1/userinfo?fields=id,email`: 

First request took 146 ms. Subsequent requests took 50 to 75 ms (mean 58), possibly indicating server-side caching.

danielsotirhos commented on src/azul/__init__.py:
```diff
@@ -820,6 +820,10 @@ def reindex_sources(self) -> List[str]:
 
+    @property
+    def google_oauth_client_id(self) -> str:
+        return os.environ['AZUL_GOOGLE_OAUTH_CLIENT_ID']
```
```suggestion
        return os.environ.get('AZUL_GOOGLE_OAUTH_CLIENT_ID')
```
Is it a requirement that all deployments have this property set?

noah-aviel-dove replied:
It's going to complicate the Swagger configuration if it's allowed to be unset, so I'll say "yes" until instructed otherwise.

dsotirho-ucsc reviewed (changes requested):
Looks good. Just a couple small things to consider.

dsotirho-ucsc reviewed (approved):
LGTM

hannes-ucsc commented on README.md:
```diff
@@ -299,8 +300,34 @@ Alternatively, create an `environment.local.py` file in the project root
+
+3. For *Application Type*, select *Web application*
+
+4. For *Name*, ``_select`` your deployment, evaluate ``azul.config.service_name`` in python, and enter the result
```
I think you can assume that the target deployment is selected (or add instructions to do so as the first step), and provide the exact python -c command line, like `python -c 'from azul import config; print(config.service_name)'`

hannes-ucsc commented on README.md:
```diff
@@ -299,8 +300,34 @@ Alternatively, create an `environment.local.py` file in the project root
+
+7. Click *Create*
+
+8. Copy the OAuth Client ID (_not_ the client secret) and insert it into your `environment.py` file:
```
Avoid "your" and use "the deployment's"

hannes-ucsc commented on src/azul/__init__.py:
```diff
@@ -814,6 +814,10 @@ def reindex_sources(self) -> List[str]:
 
+    @property
+    def google_oauth_client_id(self) -> str:
+        return os.environ['AZUL_GOOGLE_OAUTH_CLIENT_ID']
```
Needs declaration and documentation in top-level environment.py

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -107,7 +131,21 @@ def spec(self) -> JSON:
             ],
-            'servers': [{'url': self.self_url('/')}]
+            'servers': [{'url': self.self_url('/')}],
+            'security': [{self.app_name: self.oauth2_client.scopes}],
```
Doesn't this indicate that all endpoints *require* authentication? If so, that would not be correct. I think this should declare the optionality but you should investigate:

```suggestion
            'security': [{}, {self.app_name: self.oauth2_client.scopes}],
```

hannes-ucsc replied:
Also, `component` should precede `security` since the latter references the former. That's what all the examples look like.

noah-aviel-dove replied:
StackOverflow recapitulates your usage of the empty security schema and provides references to comments made by the OpenAPI developers.

hannes-ucsc commented on src/azul/oauth2.py:
```diff
@@ -0,0 +1,52 @@
+    user_info_url = 'https://www.googleapis.com/oauth2/v1/userinfo'
+
+    @cache
+    def get_auth_user(self, bearer_token: str) -> AuthUser:
```
Remove all unused functionality.

hannes-ucsc commented on test/test_openapi.py:
```diff
@@ -178,6 +185,24 @@ def route1():
 
         self.assertEqual(app.spec()['tags'], [{'name': 'foo'}, {'name': 'qux'}])
 
+    def _expected_oauth_spec(self, app_name: str):
```
This is essentially duplicating the code under test. Extract that code under test into a method/function and patch method to return "nothing" for these tests.

nadove-ucsc committed:
Fix typos in README

noah-aviel-dove commented:
The file `oauth.py` is now somewhat vestigial but I think it's good idea to keep it separate because in #3023 may expand on it.

hannes-ucsc commented on README.md:
```diff
@@ -21,14 +21,15 @@ Boardwalk, a web application for browsing genomic data sets.
-      * [3.8 Deleting all indices](#38-deleting-all-indices)
-      * [3.9 Deleting a deployment](#39-deleting-a-deployment)
+      * [3.2 Provisioning cloud infrastructure](#32-one-time-provisioning-of-deployment-specific-resources)
+      * [3.3 Provisioning cloud infrastructure](#33-automated-provisioning-of-deployment-specific-resources)
```
![image](https://user-images.githubusercontent.com/5143256/116598540-d03e9a80-a8db-11eb-888a-9047c252c0a6.png)

Duplicate title. Missing third level.

hannes-ucsc commented on README.md:
```diff
@@ -500,7 +503,36 @@ If you intend to set up a Gitlab instance for CI/CD of your Azul deployments, an
+
+3. For *Application Type*, select *Web application*
+
+4. For *Name*, enter the output from ``python3 -c 'from azul import config; print(config.service_name)'``
```
Markdown uses single back quote.

hannes-ucsc commented on README.md:
```diff
@@ -500,7 +503,36 @@ If you intend to set up a Gitlab instance for CI/CD of your Azul deployments, an
 [section on CI/CD](#9-continuous-deployment-and-integration) and for details.
 
-## 3.2 Provisioning cloud infrastructure
+## 3.2 One-time provisioning of deployment-specific resources
```
Introductory paragraph would be good here.

hannes-ucsc commented on environment.py:
```diff
@@ -404,6 +404,9 @@ def env() -> Mapping[str, Optional[str]]:
         'AZUL_SAM_SERVICE_URL': None,
 
+        # OAuth2 Client ID to be used for authenticating users.
+        'AZUL_GOOGLE_OAUTH_CLIENT_ID': None,
```
```suggestion
        'AZUL_GOOGLE_OAUTH2_CLIENT_ID': None,
```

hannes-ucsc commented on README.md:
```diff
@@ -21,14 +21,15 @@ Boardwalk, a web application for browsing genomic data sets.
-      * [3.9 Deleting a deployment](#39-deleting-a-deployment)
+      * [3.2 Provisioning cloud infrastructure](#32-one-time-provisioning-of-deployment-specific-resources)
+      * [3.3 Provisioning cloud infrastructure](#33-automated-provisioning-of-deployment-specific-resources)
+      * [3.3 Creating the Elasticsearch indices](#34-creating-the-elasticsearch-indices)
```
Anchors are off by one.

hannes-ucsc commented on src/azul/oauth2.py:
```diff
@@ -0,0 +1,4 @@
+class Oauth2Client:
```
This is anemic and should go. If sth like this is needed in a future PR, it should come in the form of a controller.

noah-aviel-dove commented:
@hannes-ucsc You didn't increment the review count or comment "no bump".

noah-aviel-dove commented:
I think we should hold on merging this for now since the upgrading instructions may change as a result of the ongoing conversation with the TDR team regarding the list allowed OAuth2 client IDs.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -40,6 +40,22 @@ def __init__(self, app_name, unit_test=False, spec=None):
 
+    def _bearer_token(self, request: Request) -> Optional[str]:
+        try:
+            auth_header = request.headers.get('Authorization') or request.headers['authorization']
```
HTTP header names are case-insensitive so HTTP client and server implementations typically use dedicated, case-insensitive mappings to represent them. That seems to be the case for Chalice.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -40,6 +40,22 @@ def __init__(self, app_name, unit_test=False, spec=None):
+                if auth_type.lower() == 'bearer':
+                    return auth_token
+                else:
+                    return None
```
We should reject requests carrying an authorization header that doesn't match our expectation. The response code should be 401.

noah-aviel-dove replied:
Raising 401 here causes a series of obtuse unit test failures stemming from here: https://github.com/DataBiosphere/azul/blob/dbee69ecf692c8268ab262b96b9d4b89f75595fb/test/indexer/test_hca_indexer.py#L1472

Seems that the indexer also uses the `Authorization` header for hmac.

hannes-ucsc replied:
That may be so in which case we need to add the appropriate conditional behavior to accommodate all valid cases. The service doesn't use HMAC and so it should let that pass. Vice versa, the indexer doesn't use bearer tokens and shouldn't let that pass. When it comes to security, leniency is usually ill advised.

hannes-ucsc reviewed (changes requested):
We need to document the provisioning of the shared OAuth2 client in the README section for one-time provisioning of cloud resources. We need to document the management of the redirect URIs for deployments that share the client. This seems obvious to me, let me know if I am missing something.

nadove-ucsc committed:
Remove README TOC (#2972)

hannes-ucsc commented on environment.py:
```diff
@@ -404,6 +404,9 @@ def env() -> Mapping[str, Optional[str]]:
         # instance configured in `AZUL_TDR_SERVICE_URL`.
         'AZUL_SAM_SERVICE_URL': None,
 
+        # OAuth2 Client ID to be used for authenticating users.
```
Needs more meat. Mention that this is shared and refer to the respective README section.

hannes-ucsc commented on lambdas/indexer/app.py:
```diff
@@ -45,6 +48,16 @@ def __init__(self):
+        auth_token = super().current_auth_token
+        if auth_token is None:
+            return None
+        elif auth_token.lower().startswith('bearer: '):
```
Is there a way to check for an HMAC auth header, rather than disallowing only one of the known unsupported ones? In security, you always want to disallow by default and only allow a specific set of supported cases. If there isn't, that should be explained in a comment.

hannes-ucsc commented on README.md:
```diff
@@ -296,11 +229,37 @@ account.
+directory and specify a global default for `GOOGLE_APPLICATION_CREDENTIALS`
 there.
 
+### 2.3.3 Google OAuth credentials
```
Now you moved this back to the wrong place. You have to realize that developers may not have a deployment yet when they're at this stage of the README.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -154,9 +181,13 @@ def _log_request(self):
+            token = self.current_auth_token
+            auth_status, auth_stmt = (('unauthenticated', '') if token is None
+                                      else ('authenticated', f' with token {token!r}'))
+            log.info(f"Received {auth_status} {context['httpMethod']} request "
```
Is there a particular reason you're not following the C guide here?

https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#logging 4th bullet

noah-aviel-dove replied:
I was extrapolating from the existing code but there's no good reason to be using f-strings here.

hannes-ucsc commented on test/service/test_request_logging.py:
```diff
@@ -34,12 +34,17 @@ def test_request_logs(self):
-                    (level, response_log)
-                ])
+            for auth_status, auth_stmt in [
+                ("unauthenticated", ""),
```
quotes

hannes-ucsc commented on README.md:
```diff
@@ -416,9 +349,35 @@ deploying to.
 3. Read all comments in `deployments/.active/environment.py` and
    `deployments/.active/environment.local.py` and make the appropriate edits.
+   
+## 2.3.5 Manual configuration of personal deployment
```
Still the wrong section. It needs to go into section 3.

hannes-ucsc commented on README.md:
```diff
@@ -416,9 +349,35 @@ deploying to.
+
+3. Under *OAuth 2.0 Client IDs*, select `azul-service-dev` and click the pencil icon to edit
+
+4. Add an entry to *Authorized JavaScript origins* and enter the output from `python3 -c 'from azul import config; print(config.service_endpoint)'`
```
Wrap list items on 80.

hannes-ucsc commented on README.md:
```diff
@@ -416,9 +349,35 @@ deploying to.
+
+8. `_refresh`
+
+## 2.5 PyCharm
```
What happened to section 2.4?

hannes-ucsc reviewed (changes requested):
Since there's been a snag with allow-listing our client ID and I am not so sure this is gonna be resolved this week, we should also add the temporary workaround to the documentation. I would prefix each of the two parts you added with a note that makes clear that the client setup isn't needed right now, which client ID to use instead and how to get the redirect URIs registered. Nothing too detailed.

noah-aviel-dove commented:
Example of log message:

```
[INFO] 2021-05-14T11:31:26.543Z f9b1c241-37a5-4a16-bacd-f571e2ed14ce Received HEAD request to '/index/projects' with parameters {"size": "1"} (unauthenticated).
[INFO] 2021-05-14T11:31:00.202Z c3ef035f-adf4-4b4e-97ee-4572d6a3e246 Received GET request to '/index/catalogs' without parameters (authenticated with bearer token 'ya29.<redacted>').
```

Except `<redacted>` contains the actual token.

hannes-ucsc commented:
Not done with this, yet.

hannes-ucsc committed:
Update PyCharm image used by `make format`

The previously used version formats

```
nonlocal foo
```

as

```
nonlocal
foo
```

which is invalid syntax.

hannes-ucsc committed:
[2/3] Add OAuth 2.0 authentication and log access tokens (#2951)

Formalize authentication of requests

hannes-ucsc committed:
[3/3] Add OAuth 2.0 authentication and log access tokens (#2951)

Rename test module

hannes-ucsc reviewed (changes requested):
Please address https://github.com/DataBiosphere/azul/pull/2999/files#r634624025, make sure that we consistently use "OAuth 2.0", "OAuth2" or "OAUTH2" instead of just "OAuth" in documentation and code, test this with our own client ID and remove the documentation of using the shared BI client.

nadove-ucsc committed:
Consistently specify OAuth 2.0 instead of OAuth

noah-aviel-dove commented:
I have deliberately preserved two occurrences of "OAuth" without the 2.0) in the README for consistency with the Google Cloud Console UI.

nadove-ucsc commented:
Pertinent log messages:
```
[INFO] 2021-05-21T04:14:03.236Z f0ed87b5-689b-4352-bc4c-e0ddf74bcffb Received GET request for '/index/catalogs', with query null and headers {"accept": "application/json", "accept-encoding": "gzip, deflate, br", "accept-language": "en-US,en;q=0.9", "authorization": "Bearer REDACTED", "cloudfront-forwarded-proto": "https", "cloudfront-is-desktop-viewer": "true", "cloudfront-is-mobile-viewer": "false", "cloudfront-is-smarttv-viewer": "false", "cloudfront-is-tablet-viewer": "false", "cloudfront-viewer-country": "US", "cookie": "_ga=GA1.2.385779753.1571420823; _ga=GA1.5.385779753.1571420823; amplitude_id_408774472b1245a7df5814f20e7484d0ucsc.edu=eyJkZXZpY2VJZCI6IjE4M2Y4ODdlLTdlZDMtNDkxMi1iMWIwLTI5NWM0NzBmNDdjZCIsInVzZXJJZCI6bnVsbCwib3B0T3V0IjpmYWxzZSwic2Vzc2lvbklkIjoxNTg1MDg4NjQyOTUxLCJsYXN0RXZlbnRUaW1lIjoxNTg1MDg5MTg0NjExLCJldmVudElkIjoyLCJpZGVudGlmeUlkIjo3LCJzZXF1ZW5jZU51bWJlciI6OX0=; experimentation_subject_id=IjA2ZTJlYzgyLWRhZmUtNGM0NC05MTA0LTIwOTk0ZGRmYzhlYSI%3D--40b9082cf3fe34ed13266b7c72821285a4f43ec3; nmstat=296a9fc7-4ef2-9ff6-6122-314e5764b036", "host": "service.nadove.dev.singlecell.gi.ucsc.edu", "referer": "https://service.nadove.dev.singlecell.gi.ucsc.edu/", "sec-ch-ua": "\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"90\", \"Google Chrome\";v=\"90\"", "sec-ch-ua-mobile": "?0", "sec-fetch-dest": "empty", "sec-fetch-mode": "cors", "sec-fetch-site": "same-origin", "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36", "via": "2.0 5ecb1870eacf5c5cd4ef9910916dfe9f.cloudfront.net (CloudFront)", "x-amz-cf-id": "SWdkGmjMhTuxzB7ylrapfzOsZjcYsh0IZihpZI6aN4C7x_oHYPpwzw==", "x-amzn-trace-id": "Root=1-60a7338b-0ab8fbe157100cf131233b58", "x-forwarded-for": "67.180.133.254, 64.252.173.153", "x-forwarded-port": "443", "x-forwarded-proto": "https"}.

[INFO] 2021-05-21T04:14:03.237Z f0ed87b5-689b-4352-bc4c-e0ddf74bcffb Authenticated request as ServiceApp.OAuth2(access_token='REDACTED')
```

With the `azul-dev` client ID (`713613812354...`)

hannes-ucsc commented on README.md:
```diff
@@ -500,7 +433,60 @@ If you intend to set up a Gitlab instance for CI/CD of your Azul deployments, an
+
+2. Navigate to *APIs & Services* -> *Credentials*
+
+3. Under *OAuth 2.0 Client IDs*, select `azul-service-dev` and click the pencil
```
```suggestion
3. Under *OAuth 2.0 Client IDs*, select `azul-dev` and click the pencil
```

hannes-ucsc commented on README.md:
```diff
@@ -464,7 +397,7 @@ module is part of a rarely used feature that can be disabled by unchecking
 are provisioned automatically by `make deploy`. A handful of  resources must be
-created manually before invoking thise Makefile targets for the first time in a
+created manually before invoking these Makefile targets for the first time in a
 particular AWS account. This only needs to be done once per AWS account, before
 the first Azul deployment is created in that account. Additional deployments do 
 not require this step.
```
This is now incorrect as we're touching the Google project, too.

hannes-ucsc commented on README.md:
```diff
@@ -480,17 +413,17 @@ the bucket, you may want to include the region name at then end of the bucket
 name. That way you can have consistent bucket names across regions.
 
 Create a Route 53 hosted zone for the Azul service and indexer. Multiple
```
Add subsection heading "Route 53 hosted zones"

hannes-ucsc commented on README.md:
```diff
@@ -500,7 +433,60 @@ If you intend to set up a Gitlab instance for CI/CD of your Azul deployments, an
 EBS volume needs to be created as well. See [gitlab.tf.json.template.py] and the
```
Add subsection heading "EBS volume for Gitlab"

hannes-ucsc commented on README.md:
```diff
@@ -500,7 +433,60 @@ If you intend to set up a Gitlab instance for CI/CD of your Azul deployments, an
 [section on CI/CD](#9-continuous-deployment-and-integration) and for details.
 
-## 3.2 Provisioning cloud infrastructure
+
```
Add subsection heading "Google OAuth 2.0 client for Terra"

hannes-ucsc reviewed (changes requested):
We usually don't revert commits in the same feature branch. Please create a clean history. I don't need to see fixups this time around.
